File: tests/translator/wvelminiapp.f90
program wvelminapp

      integer, parameter :: N = 11
      integer, parameter :: Istr = 1
      integer, parameter :: Jstr = 1
      integer, parameter :: Iend = 1000
      integer, parameter :: Jend = 1000
      integer, parameter :: r8 = 8

      integer, parameter :: LBi = Istr-1
      integer, parameter :: UBi = Iend+1
      integer, parameter :: LBj = Jstr-1
      integer, parameter :: UBj = Jend+1

      real(r8) :: pm(LBi:UBi,LBj:UBj)
      real(r8) :: pn(LBi:UBi,LBj:UBj)
      real(r8) :: z_r(LBi:UBi,LBj:UBj,N)
      real(r8) :: z_w(LBi:UBi,LBj:UBj,0:N)
      real(r8) :: u(LBi:UBi,LBj:UBj,N,2)
      real(r8) :: v(LBi:UBi,LBj:UBj,N,2)

      real(r8), dimension(Istr:Iend+1,Jstr:Jend+1,N) :: vert

      real(r8), dimension(Istr:Iend+1,Jstr:Jend+1) :: wrk

      integer :: i, j, k, NN
      integer :: Ninp, iter
      real(r8) :: res, t1, t2

      Ninp = 1


      DO NN = 1,2
        DO k=1,N
          DO j=LBj,UBj
            DO i=LBi,UBi
              u(i,j,k,NN) = i-j-k*2.0
              v(i,j,k,NN) = i-j+k*2.0
              pm(i,j) = i+1
              pn(i,j) = j+1
              z_r(i,j,k) = j-i-k
            END DO
          END DO
        END DO
      END DO

      print*, SHAPE(z_r)
          print*, SUM(z_r), SUM(u), SUM(v), SUM(pm), SUM(pn)


      call CPU_TIME(t1)
      !do iter2=1,10
          res = 0

          DO iter=1,1011

              DO k=1,N
                DO j=Jstr,Jend
                  DO i=Istr,Iend+1
                    wrk(i,j)=u(i,j,k,Ninp)*(z_r(i,j,k)-z_r(i-1,j,k))*(pm(i-1,j)+pm(i,j))

                  END DO
                  DO i=Istr,Iend
                    vert(i,j,k)=0.25_r8*(wrk(i,j)+wrk(i+1,j))
                    !print *, i, j, vert(i,j,k)

                  END DO
                END DO
                DO j=Jstr,Jend+1
                  DO i=Istr,Iend
                    wrk(i,j)=v(i,j,k,Ninp)*(z_r(i,j,k)-z_r(i,j-1,k))*(pn(i,j-1)+pn(i,j))

                  END DO
                END DO
                DO j=Jstr,Jend
                  DO i=Istr,Iend
                    vert(i,j,k)=vert(i,j,k)+0.25_r8*(wrk(i,j)+wrk(i,j+1))
                    !print *, i, j, 0.25_r8*(wrk(i,j)+wrk(i,j+1)), vert(i,j,k)
                  END DO
                END DO
              END DO
              !print *,shape(vert)
              res = res + sum(vert) ! (Istr+1:Iend,Jstr+1:Jend,:))
              !print *, res
          END DO
          print *,res
      !end do
      call CPU_TIME(t2)

      print *, "TIME: ", (t2-t1)/10

end program wvelminapp

File: tests/translator/wvelminiapp.py
import cupy as cp
from cupyx.time import repeat

N = 11
Istr = 1
Jstr = 1
Iend = 1000
Jend = 1000
r8 = 8

LBi = Istr - 1
UBi = Iend + 1
LBj = Jstr - 1
UBj = Jend + 1

pm = cp.zeros((UBj - LBj + 1, UBi - LBi + 1), dtype=cp.float64)
pn = cp.zeros((UBj - LBj + 1, UBi - LBi + 1), dtype=cp.float64)
z_r = cp.zeros((N, UBj - LBj + 1, UBi - LBi + 1), dtype=cp.float64)
u = cp.zeros((2, N, UBj - LBj + 1, UBi - LBi + 1), dtype=cp.float64)
v = cp.zeros((2, N, UBj - LBj + 1, UBi - LBi + 1), dtype=cp.float64)

vert = cp.zeros((N, UBj - LBj, UBi - LBi ), dtype=cp.float64)
wrk = cp.zeros((N, UBj - LBj , UBi - LBi ), dtype=cp.float64)

Ninp = 1
res = 0

# Initialization
k, j, i = cp.indices(z_r.shape)
print(z_r.shape)
for l in range(2):
    u[l, :, :, :] = i - j - k * 2.0 - 2
    v[l, :, :, :] = i - j + k * 2.0 + 2
    pm[:, :] = i[0, :, :] + 1
    pn[:, :] = j[0, :, :] + 1
    z_r[:, :, :] = j - i - k - 1
print(z_r.sum(), u.sum(), v.sum(), pm.sum(), pn.sum())

def trivialCupy():
    res = 0.0
    for iter in range(10):

        wrk[:, :-1, :] = u[Ninp, :, 1:-1, 1:] * (z_r[:, 1:-1, 1:] - z_r[:, 1:-1, :-1]) * cp.tile(pm[1:-1, 0:-1] + pm[1:-1, 1:], [N,1,1])
        vert[:,:,1:] = 0.25*(wrk[:,:, 1:] + wrk[:,:,:-1])

        wrk[:, :, :-1] = v[Ninp, :, 1:, 1:-1] * (z_r[:, 1:, 1:-1] - z_r[:, :-1, 1:-1]) * cp.tile(pn[0:-1, 1:-1] + pn[1:, 1:-1], [N,1,1])
        vert[:, :-1, 1:] += 0.25 * (wrk[:, 1:, :-1] + wrk[:, :-1, :-1])

        res += vert[:,:-1,:].sum(axis=None)
        print(iter, res)

@cp.fuse()
def fuseRtoU(u1, u2):
    return 0.25*(u1 + u2)


@cp.fuse()
def fuseRtoV(u1, u2):
    return 0.25*(u1 + u2)

@cp.fuse()
def fuseDηRtoU(u1, u2):
    return (u1 - u2)

@cp.fuse()
def fuseDyRtoV(u1, u2):
    return (u1 - u2)



def fuseAndCupy():
    res = 0.0
    for iter in range(10):
        wrk[:, :-1, :] = u[Ninp, :, 1:-1, 1:] * fuseDηRtoU(z_r[:, 1:-1, 1:], z_r[:, 1:-1, :-1]) * (pm[None,1:-1, 0:-1] + pm[None,1:-1, 1:])
        vert[:,:,1:] = fuseRtoU(wrk[:,:, 1:], wrk[:,:,:-1])

        wrk[:, :, :-1] = v[Ninp, :, 1:, 1:-1] * fuseDyRtoV(z_r[:, 1:, 1:-1], z_r[:, :-1, 1:-1]) * (pn[None, 0:-1, 1:-1] + pn[None, 1:, 1:-1])
        vert[:, :-1, 1:] += fuseRtoV(wrk[:, 1:, :-1] , wrk[:, :-1, :-1])

        # res += vert[:,:-1,:].sum(axis=None)
    print(iter, res)

kernelfunc = cp.ElementwiseKernel(
    'float64 z_r2d, raw float64 z_r, raw float64 u, raw float64 v, raw float64 pn, raw float64 pm',
    'float64 vert',
'''
    double wrk;
    wrk = u[i]*(z_r[i] - z_r[i-1])*(pm[i-1]+pm[i])
//   printf("%i ", i)    ;
// _ind.size()    ;
//  vert=1.0
''',
    'kernelfunc', options=('--dopt=on', '--use_fast_math'))
# wrk[:, :-1, :] = u[Ninp, :, 1:-1, 1:] * (z_r[:, 1:-1, 1:] - z_r[:, 1:-1, :-1]) * cp.tile(pm[1:-1, 0:-1] + pm[1:-1, 1:], [N,1,1])
#         vert[:,:,1:] = 0.25*(wrk[:,:, 1:] + wrk[:,:,:-1])
#
#         wrk[:, :, :-1] = v[Ninp, :, 1:, 1:-1] * (z_r[:, 1:, 1:-1] - z_r[:, :-1, 1:-1]) * cp.tile(pn[0:-1, 1:-1] + pn[1:, 1:-1], [N,1,1])
#         vert[:, :-1, 1:] += 0.25 * (wrk[:, 1:, :-1] + wrk[:, :-1, :-1])
#
#         res += vert[:,:-1,:].sum(axis=None)


# def kernelCupy():
#     res = 0.0
#     for iter in range(100):
#
#         # vert = kernelfunc(z_r, z_r, u[Ninp,:,:,:], v[Ninp,:,:,:], pn, pm)
#         vert = fuseCupy(z_r)
#
# kernelCupy()
print(repeat(trivialCupy, (), n_repeat=10, n_warmup=10))
print(repeat(fuseAndCupy, (), n_repeat=10, n_warmup=10))

